File: src/android/com/keyri/cordova/plugin/CordovaKeyri.kt
package com.keyri.cordova.plugin

import android.net.Uri
import android.util.Log
import android.app.Activity
import android.content.Intent
import androidx.fragment.app.FragmentActivity
import com.keyrico.keyrisdk.Keyri
import com.keyrico.keyrisdk.entity.session.Session
import com.keyrico.keyrisdk.sec.fraud.enums.EventType
import com.google.gson.Gson
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.json.JSONArray
import org.json.JSONObject
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import com.keyrico.scanner.easyKeyriAuth

class CordovaKeyri : CordovaPlugin() {

    private var activeSession: Session? = null

    private lateinit var keyri: Keyri
    private lateinit var appKey: String
    private var publicApiKey: String? = null
    private var serviceEncryptionKey: String? = null
    private var blockEmulatorDetection: Boolean = true

    private var easyKeyriAuthCallback: CallbackContext? = null

    override fun execute(action: String, arguments: JSONArray?, callbackContext: CallbackContext): Boolean {
        Log.e("Executing native method", "$action, $arguments")

        when (action) {
            "initialize" -> {
                val appKey = arguments?.getString(0)
                val publicApiKey = arguments?.getString(1)
                val serviceEncryptionKey = arguments?.getString(2)
                val blockEmulatorDetection = arguments?.getBoolean(3)

                initialize(appKey, publicApiKey, serviceEncryptionKey, blockEmulatorDetection ?: true, callbackContext)
            }

            "isInitialized" -> {
                if (this::keyri.isInitialized) {
                    callbackContext.success()
                } else {
                    callbackContext.error("isInitialized, Keyri is not initialized")
                }
            }

            "easyKeyriAuth" -> {
                val payload = arguments?.getString(0)
                val publicUserId = arguments?.getString(1)

                easyKeyriAuth(payload, publicUserId, callbackContext)
            }

            "generateAssociationKey" -> {
                val publicUserId = arguments?.getString(0)

                generateAssociationKey(publicUserId, callbackContext)
            }

            "generateUserSignature" -> {
                val publicUserId = arguments?.getString(0)
                val data = arguments?.getString(1)

                generateUserSignature(publicUserId, data, callbackContext)
            }

            "listAssociationKeys" -> listAssociationKeys(callbackContext)
            "listUniqueAccounts" -> listUniqueAccounts(callbackContext)
            "getAssociationKey" -> {
                val publicUserId = arguments?.getString(0)

                getAssociationKey(publicUserId, callbackContext)
            }

            "removeAssociationKey" -> {
                val publicUserId = arguments?.getString(0)

                removeAssociationKey(publicUserId, callbackContext)
            }

            "sendEvent" -> {
                val publicUserId = arguments?.getString(0)
                val eventType = arguments?.getString(1)
                val success = arguments?.getBoolean(2) ?: true

                sendEvent(publicUserId, eventType, success, callbackContext)
            }

            "initiateQrSession" -> {
                val sessionId = arguments?.getString(0)
                val publicUserId = arguments?.getString(1)

                initiateQrSession(sessionId, publicUserId, callbackContext)
            }

            "login" -> {
                val publicUserId = arguments?.getString(0)

                login(publicUserId, callbackContext)
            }

            "register" -> {
                val publicUserId = arguments?.getString(0)

                register(publicUserId, callbackContext)
            }

            "initializeDefaultConfirmationScreen" -> {
                val payload = arguments?.getString(1)

                initializeDefaultConfirmationScreen(payload, callbackContext)
            }

            "processLink" -> {
                val link = arguments?.getString(0)
                val payload = arguments?.getString(1)
                val publicUserId = arguments?.getString(2)

                processLink(link, payload, publicUserId, callbackContext)
            }

            "confirmSession" -> {
                val payload = arguments?.getString(0)
                val trustNewBrowser = arguments?.getBoolean(1) ?: false

                confirmSession(payload, trustNewBrowser, callbackContext)
            }

            "denySession" -> {
                val payload = arguments?.getString(0)

                denySession(payload, callbackContext)
            }

            else -> {
                callbackContext.error("Action $action not implemented")

                return false
            }
        }

        return true
    }

    private fun initialize(
        appKey: String?,
        publicApiKey: String?,
        serviceEncryptionKey: String?,
        blockEmulatorDetection: Boolean,
        callback: CallbackContext
    ) {
        cordova.threadPool.execute {
            try {
                if (appKey == null) {
                    callback.error("initialize, appKey must not be null")
                } else {
                    cordova.getActivity()?.let {
                        if (!this::keyri.isInitialized) {
                            this.appKey = appKey
                            this.publicApiKey = publicApiKey
                            this.serviceEncryptionKey = serviceEncryptionKey
                            this.blockEmulatorDetection = blockEmulatorDetection

                            keyri = Keyri(it, appKey, publicApiKey, serviceEncryptionKey, blockEmulatorDetection)
                        }

                        callback.success()
                    }
                }
            } catch (e: Exception) {
                callback.error(e.message)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == AUTH_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                easyKeyriAuthCallback?.success()
            }
        }
    }

    private fun easyKeyriAuth(
        payload: String?,
        publicUserId: String?,
        callback: CallbackContext
    ) {
        if (payload == null) {
            callback.error("easyKeyriAuth, payload must not be null")
        } else {
            cordova.getActivity()?.let {
                cordova.setActivityResultCallback(this)
                easyKeyriAuth(it, AUTH_REQUEST_CODE, appKey, publicApiKey, serviceEncryptionKey, blockEmulatorDetection, payload, publicUserId)

                easyKeyriAuthCallback = callback
            } ?: callback.error("initializeDefaultScreen, can't get cordova.getActivity()")
        }
    }

    private fun generateAssociationKey(publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            val associationKey = publicUserId?.let {
                keyri.generateAssociationKey(publicUserId).getOrThrow()
            } ?: keyri.generateAssociationKey().getOrThrow()

            callback.success(associationKey)
        }
    }

    private fun generateUserSignature(publicUserId: String?, data: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (data == null) {
                callback.error("generateUserSignature, data must not be null")
            } else {
                val userSignature = publicUserId?.let {
                    keyri.generateUserSignature(it, data).getOrThrow()
                } ?: keyri.generateUserSignature(data = data).getOrThrow()

                callback.success(userSignature)
            }
        }
    }

    private fun listAssociationKeys(callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            val keys = keyri.listAssociationKeys().getOrThrow()
            val resultArray = JSONArray()

            keys.map {
                val keyObject = JSONObject().apply {
                    put(it.key, it.value)
                }

                resultArray.put(keyObject)
            }

            callback.success(resultArray)
        }
    }

    private fun listUniqueAccounts(callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            val keys = keyri.listUniqueAccounts().getOrThrow()
            val resultArray = JSONArray()

            keys.map {
                val keyObject = JSONObject().apply {
                    put(it.key, it.value)
                }

                resultArray.put(keyObject)
            }

            callback.success(resultArray)
        }
    }

    private fun getAssociationKey(publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            val associationKey = publicUserId?.let {
                keyri.getAssociationKey(publicUserId).getOrThrow()
            } ?: keyri.getAssociationKey().getOrThrow()

            callback.success(associationKey)
        }
    }

    private fun removeAssociationKey(publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (publicUserId == null) {
                callback.error("removeAssociationKey, publicUserId must not be null")
            } else {
                keyri.removeAssociationKey(publicUserId).getOrThrow()
                callback.success()
            }
        }
    }

    private fun sendEvent(publicUserId: String?, eventType: String?, success: Boolean, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            val type = EventType.values().firstOrNull { it.type == eventType }

            if (type == null) {
                callback.error("sendEvent, eventType must not be null")
            } else {
                val userId = publicUserId ?: "ANON"

                keyri.sendEvent(userId, type, success).onSuccess {
                    callback.success()
                }.onFailure {
                    callback.error("sendEvent, ${it.message}")
                }
            }
        }
    }

    private fun initiateQrSession(sessionId: String?, publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (sessionId == null) {
                callback.error("initiateQrSession, sessionId must not be null")
            } else {
                keyri.initiateQrSession(sessionId, publicUserId).onSuccess { session ->
                    activeSession = session

                    callback.success(JSONObject(Gson().toJson(session)))
                }.onFailure {
                    callback.error("initiateQrSession, ${it.message}")
                }
            }
        }
    }

    private fun login(publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            keyri.login(publicUserId).onSuccess { loginObject ->
                callback.success(JSONObject(Gson().toJson(loginObject)))
            }.onFailure {
                callback.error("login, ${it.message}")
            }
        }
    }

    private fun register(publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            keyri.register(publicUserId).onSuccess { registerObject ->
                callback.success(JSONObject(Gson().toJson(registerObject)))
            }.onFailure {
                callback.error("register, ${it.message}")
            }
        }
    }

    private fun initializeDefaultConfirmationScreen(payload: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (activeSession == null) {
                callback.error("initializeDefaultConfirmationScreen, can't find session")
            } else if (payload == null) {
                callback.error("initializeDefaultConfirmationScreen, payload must not be null")
            } else {
                (cordova.getActivity() as? FragmentActivity)?.supportFragmentManager?.let { fm ->
                    keyri.initializeDefaultConfirmationScreen(fm, requireNotNull(activeSession), payload).onSuccess { authResult ->
                        callback.success()
                    }.onFailure {
                        callback.error("initializeDefaultConfirmationScreen, ${it.message}")
                    }
                } ?: callback.error("initializeDefaultConfirmationScreen, can't get supportFragmentManager")
            }
        }
    }

    private fun processLink(link: String?, payload: String?, publicUserId: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (link == null) {
                callback.error("processLink, link must not be null")
            } else if (payload == null) {
                callback.error("processLink, payload must not be null")
            } else {
                (cordova.getActivity() as? FragmentActivity)?.supportFragmentManager?.let { fm ->
                    keyri.processLink(fm, Uri.parse(link), payload, publicUserId).onSuccess { authResult ->
                        callback.success()
                    }.onFailure {
                        callback.error("processLink, ${it.message}")
                    }
                }
            }
        }
    }

    private fun confirmSession(payload: String?, trustNewBrowser: Boolean, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (activeSession == null) {
                callback.error("confirmSession, can't find session")
            } else if (payload == null) {
                callback.error("confirmSession, payload must not be null")
            } else {
                activeSession?.confirm(payload, requireNotNull(cordova.getActivity()), trustNewBrowser)?.onSuccess {
                    callback.success()
                }?.onFailure {
                    callback.error("confirmSession, ${it.message}")
                }
            }
        }
    }

    private fun denySession(payload: String?, callback: CallbackContext) {
        keyriCoroutineScope(callback).launch {
            if (activeSession == null) {
                callback.error("denySession, can't find session")
            } else if (payload == null) {
                callback.error("denySession, payload must not be null")
            } else {
                activeSession?.deny(payload, requireNotNull(cordova.getActivity()))?.onSuccess {
                    callback.success()
                }?.onFailure {
                    callback.error("denySession, ${it.message}")
                }
            }
        }
    }

    private fun keyriCoroutineScope(callback: CallbackContext): CoroutineScope {
        val exceptionHandler = CoroutineExceptionHandler { _, e ->
            callback.error(e.message ?: "Error calling this method")
        }

        return CoroutineScope(Dispatchers.IO + exceptionHandler)
    }

    companion object {
        private const val AUTH_REQUEST_CODE = 2133
    }
}

File: src/iOS/com/keyri/cordova/plugin/CordovaKeyri.swift
import Foundation
import Keyri

@objc(CordovaKeyri) class CordovaKeyri : CDVPlugin {
    var keyri: KeyriInterface?
    var activeSession: Session?

    @objc(initialize:)
    func initialize(command: CDVInvokedUrlCommand) {
        guard let arguments = command.arguments else {
            processError(error: "initialize, no arguments provided", command: command)
            return
        }

        if arguments.count < 1 || !(arguments[0] is String) {
            processError(error: "initialize, first parameter must be a String", command: command)
            return
        }

        let appKey = arguments[0] as! String
        let publicApiKey = arguments[1] as? String
        let serviceEncryptionKey = arguments[2] as? String
        let blockEmulatorDetection = arguments[3] as? Bool ?? true

        keyri = KeyriInterface(appKey: appKey, publicApiKey: publicApiKey, serviceEncryptionKey: serviceEncryptionKey, blockEmulatorDetection: blockEmulatorDetection)

        processResult(command: command)
    }

    @objc(isInitialized:)
    func isInitialized(command: CDVInvokedUrlCommand) {
        if (keyri != nil) {
            processResult(command: command)
        } else {
            sendIsNotInitialized(methodName: "isInitialized", command: command)
        }
    }

    @objc(easyKeyriAuth:)
    func easyKeyriAuth(command: CDVInvokedUrlCommand) {
        guard let payload = command.arguments[0] as? String else {
            processError(error: "easyKeyriAuth, payload must not be nil", command: command)
            return
        }

        let publicUserId = command.arguments[1] as? String

        keyri?.easyKeyriAuth(payload: payload, publicUserId: publicUserId) { result in
            switch result {
            case .success:
                self.processResult(command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "easyKeyriAuth", command: command)
    }

    @objc(generateAssociationKey:)
    func generateAssociationKey(command: CDVInvokedUrlCommand) {
        guard let publicUserId = command.arguments[0] as? String else {
            processError(error: "generateAssociationKey, publicUserId must not be nil", command: command)
            return
        }

        keyri?.generateAssociationKey(publicUserId: publicUserId) { result in
            switch result {
            case .success(let key):
                self.processResult(message: key.rawRepresentation.base64EncodedString(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "generateAssociationKey", command: command)
    }

    @objc(generateUserSignature:)
    func generateUserSignature(command: CDVInvokedUrlCommand) {
        guard let publicUserId = command.arguments[0] as? String else {
            processError(error: "generateUserSignature, publicUserId must not be nil", command: command)
            return
        }

        guard let dataStr = command.arguments[1] as? String,
              let data = dataStr.data(using: .utf8) else {
            processError(error: "generateUserSignature, data must not be nil", command: command)
            return
        }

        keyri?.generateUserSignature(publicUserId: publicUserId, data: data) { result in
            switch result {
            case .success(let signature):
                self.processResult(message: signature.derRepresentation.base64EncodedString(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "generateUserSignature", command: command)
    }

    @objc(listAssociationKeys:)
    func listAssociationKeys(command: CDVInvokedUrlCommand) {
        keyri?.listAssociationKeys() { result in
            switch result {
            case .success(let list):
                self.processResult(message: list ?? [:], command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "listAssociationKeys", command: command)
    }

    @objc(listUniqueAccounts:)
    func listUniqueAccounts(command: CDVInvokedUrlCommand) {
        keyri?.listUniqueAccounts() { result in
            switch result {
            case .success(let list):
                self.processResult(message: list ?? [:], command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "listUniqueAccounts", command: command)
    }

    @objc(getAssociationKey:)
    func getAssociationKey(command: CDVInvokedUrlCommand) {
        let publicUserId = command.arguments[0] as? String

        keyri?.getAssociationKey(publicUserId: publicUserId ?? "") { result in
            switch result {
            case .success(let key):
                self.processResult(message: key?.rawRepresentation.base64EncodedString(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "getAssociationKey", command: command)
    }

    @objc(removeAssociationKey:)
    func removeAssociationKey(command: CDVInvokedUrlCommand) {
        guard let publicUserId = command.arguments[0] as? String else {
            processError(error: "removeAssociationKey, publicUserId must not be nil", command: command)
            return
        }

        keyri?.removeAssociationKey(publicUserId: publicUserId) { result in
            switch result {
            case .success:
                self.processResult(command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "removeAssociationKey", command: command)
    }

    @objc(sendEvent:)
    func sendEvent(command: CDVInvokedUrlCommand) {
        guard let publicUserId = command.arguments[0] as? String else {
            processError(error: "sendEvent, publicUserId must not be nil", command: command)
            return
        }

        guard let eventType = command.arguments[1] as? String else {
                processError(error: "sendEvent, eventType must not be nil", command: command)
                return
            }

        let eventMetadata = command.arguments[2] as? String
        let success = command.arguments[3] as? String ?? "true"

        keyri?.sendEvent(publicUserId: publicUserId, eventType: EventType.custom(name: eventType, metadata: eventMetadata), success: Bool(success) ?? true) { result in
            switch result {
            case .success(let fingerprintResponse):
                self.processResult(message: fingerprintResponse.asDictionary(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "sendEvent", command: command)
    }

    @objc(createFingerprint)
    func createFingerprint(command: CDVInvokedUrlCommand) {
        keyri?.createFingerprint() { result in
            switch result {
            case .success(let fingerprintRequest):
                self.processResult(message: fingerprintRequest.asDictionary(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "createFingerprint", command: command)
    }

    @objc(initiateQrSession:)
    func initiateQrSession(command: CDVInvokedUrlCommand) {
        guard let sessionId = command.arguments[0] as? String else {
            processError(error: "initiateQrSession, sessionId must not be nil", command: command)
            return
        }

        let publicUserId = command.arguments[1] as? String

        keyri?.initiateQrSession(sessionId: sessionId, publicUserId: publicUserId) { result in
            switch result {
            case .success(let session):
                self.activeSession = session
                self.processResult(message: session.asDictionary(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "initiateQrSession", command: command)
    }

    @objc(login:)
    func login(command: CDVInvokedUrlCommand) {
        let publicUserId = command.arguments[1] as? String

        keyri?.login(publicUserId: publicUserId) { result in
            switch result {
            case .success(let loginObject):
                self.processResult(message: loginObject.asDictionary(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "login", command: command)
    }

    @objc(register:)
    func register(command: CDVInvokedUrlCommand) {
        let publicUserId = command.arguments[1] as? String

        keyri?.register(publicUserId: publicUserId) { result in
            switch result {
            case .success(let registerObject):
                self.processResult(message: registerObject.asDictionary(), command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "register", command: command)
    }

    @objc(initializeDefaultConfirmationScreen:)
    func initializeDefaultConfirmationScreen(command: CDVInvokedUrlCommand) {
        guard let session = activeSession else {
            processError(error: "initializeDefaultConfirmationScreen, activeSession must not be nil", command: command)
            return
        }

        guard let payload = command.arguments[0] as? String else {
            processError(error: "initializeDefaultConfirmationScreen, payload must not be nil", command: command)
            return
        }

        keyri?.initializeDefaultConfirmationScreen(session: session, payload: payload) { result in
            switch result {
            case .success:
                self.processResult(command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "initializeDefaultConfirmationScreen", command: command)
    }

    @objc(processLink:)
    func processLink(command: CDVInvokedUrlCommand) {
        guard let link = command.arguments[0] as? String,
          let url = URL(string: link) else {
            processError(error: "processLink, link must not be nil", command: command)
            return
        }

        guard let payload = command.arguments[1] as? String else {
            processError(error: "processLink, payload must not be nil", command: command)
            return
        }

        guard let publicUserId = command.arguments[2] as? String else {
            processError(error: "processLink, publicUserId must not be nil", command: command)
            return
        }

        keyri?.processLink(url: url, payload: payload, publicUserId: publicUserId) { result in
            switch result {
            case .success:
                self.processResult(command: command)
            case .failure(let error):
                self.processError(error: error, command: command)
            }
        } ?? sendIsNotInitialized(methodName: "processLink", command: command)
    }

    @objc(confirmSession:)
    func confirmSession(command: CDVInvokedUrlCommand) {
        guard let payload = command.arguments[0] as? String,
            let trustNewBrowserString = command.arguments[1] as? String,
            let trustNewBrowser = Bool(trustNewBrowserString) else {
            processError(error: "confirmSession, payload must not be nil", command: command)
            return
        }

        activeSession?.confirm(payload: payload, trustNewBrowser: trustNewBrowser) { result in
            switch result {
            case .some(let error):
                self.processError(error: error, command: command)
            case .none:
                self.processResult(command: command)
            }
        } ?? processError(error: "confirmSession, activeSession is nil", command: command)
    }

    @objc(denySession:)
    func denySession(command: CDVInvokedUrlCommand) {
        guard let payload = command.arguments[0] as? String else {
            processError(error: "denySession, payload must not be nil", command: command)
            return
        }

        activeSession?.deny(payload: payload) { result in
            switch result {
            case .some(let error):
                self.processError(error: error, command: command)
            case .none:
                self.processResult(command: command)
            }
        } ?? processError(error: "denySession, activeSession is nil", command: command)
    }

    private func processResult(message: String?, command: CDVInvokedUrlCommand) {
        commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message), callbackId: command.callbackId)
    }

    private func processResult(message: [String : String]?, command: CDVInvokedUrlCommand) {
        commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message), callbackId: command.callbackId)
    }

    private func processResult(command: CDVInvokedUrlCommand) {
        commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
    }

    private func processError(error: String, command: CDVInvokedUrlCommand) {
        commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: error), callbackId: command.callbackId)
    }

    private func sendIsNotInitialized(methodName: String, command: CDVInvokedUrlCommand) {
        processError(error: methodName + ", Keyri is not initialized", command: command)
    }

    private func processError(error: Error, command: CDVInvokedUrlCommand) {
        processError(error: error.localizedDescription, command: command)
    }
}

extension Encodable {
    func asDictionary() -> [String: String]? {
        guard let data = try? JSONEncoder().encode(self) else {
            return nil
        }
        guard let dictionary = try? JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: String] else {
            return nil
        }
        return dictionary
    }
}
